test/extended/apiserver/apply.go

```diff
@@ -0,0 +1,168 @@
+		}
+
+		g.It(fmt.Sprintf("should work for %s", gvr), func() {
+			// Create at most one namespace only if needed.
+			var testNamespace string
+
+			for _, prerequisite := range data.Prerequisites {
+				resourceClient, unstructuredObj, namespace := createResource(oc, mapper, prerequisite.GvrData, prerequisite.Stub, testNamespace)
+				testNamespace = namespace
+				defer deleteResource(resourceClient, unstructuredObj.GetName())
+			}
+
+			resourceClient, unstructuredObj, namespace := createResource(oc, mapper, gvr, data.Stub, testNamespace)
+			testNamespace = namespace
+			defer deleteResource(resourceClient, unstructuredObj.GetName())
+
+			serializedObj, err := json.Marshal(unstructuredObj.Object)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By(fmt.Sprintf("updating the %s via apply", unstructuredObj.GetKind()))
+			obj, err := resourceClient.Patch(
+				context.Background(),
+				unstructuredObj.GetName(),
+				types.ApplyPatchType,
+				serializedObj,
+				metav1.PatchOptions{
+					FieldManager: "apply_test",
+				},
+			)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By(fmt.Sprintf("checking that the field managers are as expected"))
+			accessor, err := meta.Accessor(obj)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			managedFields := accessor.GetManagedFields()
+			o.Expect(managedFields).NotTo(o.BeNil())
+			if !findManager(managedFields, "create_test") {
+				g.Fail(fmt.Sprintf("Couldn't find create_test: %v", managedFields))
+			}
+			if !findManager(managedFields, "apply_test") {
+				g.Fail(fmt.Sprintf("Couldn't find apply_test: %v", managedFields))
+			}
+		})
+	}
+})
+
+func findManager(managedFields []metav1.ManagedFieldsEntry, manager string) bool {
+	for _, entry := range managedFields {
+		if entry.Manager == manager {
+			return true
+		}
+	}
+	return false
+}
+
+func deleteResource(resourceClient dynamic.ResourceInterface, name string) {
+	err := resourceClient.Delete(context.Background(), name, metav1.DeleteOptions{})
+	if err != nil {
+		framework.Logf("Unexpected error deleting resource: %v", err)
+	}
+}
+
+func createResource(
+	oc *exutil.CLI,
+	mapper *restmapper.DeferredDiscoveryRESTMapper,
+	gvr schema.GroupVersionResource,
+	stub, testNamespace string) (
+	dynamic.ResourceInterface, *unstructured.Unstructured, string) {
+
+	// Discover the gvk from the gvr
+	gvk, err := mapper.KindFor(gvr)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	// Supply a value for namespace if the scope requires
+	mapping, err := mapper.RESTMapping(gvk.GroupKind())
+	o.Expect(err).NotTo(o.HaveOccurred())
+	namespace := ""
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace && len(testNamespace) == 0 {
+		// Create the namespace on demand
+		namespace = oc.SetupProject()
+		testNamespace = namespace
+	}
+
+	// Ensure that any stub embedding the etcd test namespace
+	// is updated to use local test namespace instead.
+	if len(testNamespace) > 0 {
+		stub = strings.Replace(stub, exetcd.TestNamespace, testNamespace, -1)
```

Comment:
is there a need to use the etcd storage test namespace? Would rather use a unique namespace here to avoid conflicts, and to reduce shared logic.

Reply:
I am using a unique namespace - that's `testNamespace`. It only needs to be used when the resource itself is namespaced since many of the types are cluster-scoped. This minimizes the need to provision new namespaces since that is not cheap.